### ISSUE TITLE
openssl 3.0.7 unable to establish connection to TLS 1.0 and 1.1 Apache service with openssl 1.1.1l

### ISSUE THREAD

**Issue by mazer1310** [openssl/openssl#20039]
I'm in the process of migrating from OpenSSL 1.1.1 to 3.0.7 and we have tests that connect to an Apache service where four port numbers (9010, 9011, 9012, and 9013) are configured to only accept a single protocol, TLS 1.0 for 9010, TLS 1.1 for 9011, TLS 1.2 for 9012, and TLS 1.3 for 9013.  The Apache service is 2.4.53 with mod_ssl;OpenSSL 1.1.1l

A sample of the Apache conf file would look like this (with suitable VirtualHost sections updating SSLProtocol and SSLCipherSuite for each of the other three ports, 9011, 9012, and 9013):
```
<VirtualHost *:9010>
	SSLEngine On
	SSLCertificateFile conf/ssl/ses-server.crt
	SSLCertificateKeyFile conf/ssl/ses-server.key

	SSLHonorCipherOrder		on 
	SSLCompression          off
	SSLSessionTickets       off

	SSLProtocol             TLSv1
	SSLCipherSuite			ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:DHE-RSA-AES128-SHA:AES256-SHA:AES128-SHA
</VirtualHost>
```

With the client running OpenSSL version 1.1.1, connections are successful.

However, after migrating the client to OpenSSL version 3.0.7, connections to 9010 (TLS 1.0) and 9011 (TLS 1.1) both fail.

I ran the openssl s_client -trace <server>:9010 and found that the failure appears to be occurring during key exchange with an internal error 603A0000:error:0A0C0103:SSL routines:tls_process_key_exchange:internal error:ssl\statem\statem_clnt.c:2254:
The error is identical when connecting to <server>:9011.
connections to <server>:9012 and <server>:9013 both succeed.

Connections to all four ports are successful when connecting with openssl s_client from 1.1.1

```
CONNECTED(000001B4)
Sent Record
Header:
  Version = TLS 1.0 (0x301)
  Content Type = Handshake (22)
  Length = 323
    ClientHello, Length=319
      client_version=0x303 (TLS 1.2)
      Random:
        gmt_unix_time=0x83F8E4D5
        random_bytes (len=28): 0566F2EBB8DD10A3943BAA6AE9A077FC49C7C4C2CC91CCCE7DE30D31
      session_id (len=32): 55C19CEF92ECBD39F5B8C13E132333E898D4590E771990BAC1A3F52F91DDE3DE
      cipher_suites (len=62)
        {0x13, 0x02} TLS_AES_256_GCM_SHA384
        {0x13, 0x03} TLS_CHACHA20_POLY1305_SHA256
        {0x13, 0x01} TLS_AES_128_GCM_SHA256
        {0xC0, 0x2C} TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
        {0xC0, 0x30} TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
        {0x00, 0x9F} TLS_DHE_RSA_WITH_AES_256_GCM_SHA384
        {0xCC, 0xA9} TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
        {0xCC, 0xA8} TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
        {0xCC, 0xAA} TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256
        {0xC0, 0x2B} TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
        {0xC0, 0x2F} TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
        {0x00, 0x9E} TLS_DHE_RSA_WITH_AES_128_GCM_SHA256
        {0xC0, 0x24} TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384
        {0xC0, 0x28} TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384
        {0x00, 0x6B} TLS_DHE_RSA_WITH_AES_256_CBC_SHA256
        {0xC0, 0x23} TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256
        {0xC0, 0x27} TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
        {0x00, 0x67} TLS_DHE_RSA_WITH_AES_128_CBC_SHA256
        {0xC0, 0x0A} TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA
        {0xC0, 0x14} TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
        {0x00, 0x39} TLS_DHE_RSA_WITH_AES_256_CBC_SHA
        {0xC0, 0x09} TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
        {0xC0, 0x13} TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
        {0x00, 0x33} TLS_DHE_RSA_WITH_AES_128_CBC_SHA
        {0x00, 0x9D} TLS_RSA_WITH_AES_256_GCM_SHA384
        {0x00, 0x9C} TLS_RSA_WITH_AES_128_GCM_SHA256
        {0x00, 0x3D} TLS_RSA_WITH_AES_256_CBC_SHA256
        {0x00, 0x3C} TLS_RSA_WITH_AES_128_CBC_SHA256
        {0x00, 0x35} TLS_RSA_WITH_AES_256_CBC_SHA
        {0x00, 0x2F} TLS_RSA_WITH_AES_128_CBC_SHA
        {0x00, 0xFF} TLS_EMPTY_RENEGOTIATION_INFO_SCSV
      compression_methods (len=1)
        No Compression (0x00)
      extensions, length = 184
        extension_type=server_name(0), length=27
          0000 - 00 19 00 00 16 76 6a 61-6d 62 61 32 32 2e 64   .....vjamba22.d
          000f - 73 6f 6e 65 2e 33 64 73-2e 63 6f 6d            sone.3ds.com
        extension_type=ec_point_formats(11), length=4
          uncompressed (0)
          ansiX962_compressed_prime (1)
          ansiX962_compressed_char2 (2)
        extension_type=supported_groups(10), length=22
          ecdh_x25519 (29)
          secp256r1 (P-256) (23)
          ecdh_x448 (30)
          secp521r1 (P-521) (25)
          secp384r1 (P-384) (24)
          ffdhe2048 (256)
          ffdhe3072 (257)
          ffdhe4096 (258)
          ffdhe6144 (259)
          ffdhe8192 (260)
        extension_type=session_ticket(35), length=0
        extension_type=encrypt_then_mac(22), length=0
        extension_type=extended_master_secret(23), length=0
        extension_type=signature_algorithms(13), length=42
          ecdsa_secp256r1_sha256 (0x0403)
          ecdsa_secp384r1_sha384 (0x0503)
          ecdsa_secp521r1_sha512 (0x0603)
          ed25519 (0x0807)
          ed448 (0x0808)
          rsa_pss_pss_sha256 (0x0809)
          rsa_pss_pss_sha384 (0x080a)
          rsa_pss_pss_sha512 (0x080b)
          rsa_pss_rsae_sha256 (0x0804)
          rsa_pss_rsae_sha384 (0x0805)
          rsa_pss_rsae_sha512 (0x0806)
          rsa_pkcs1_sha256 (0x0401)
          rsa_pkcs1_sha384 (0x0501)
          rsa_pkcs1_sha512 (0x0601)
          ecdsa_sha224 (0x0303)
          rsa_pkcs1_sha224 (0x0301)
          dsa_sha224 (0x0302)
          dsa_sha256 (0x0402)
          dsa_sha384 (0x0502)
          dsa_sha512 (0x0602)
        extension_type=supported_versions(43), length=9
          TLS 1.3 (772)
          TLS 1.2 (771)
          TLS 1.1 (770)
          TLS 1.0 (769)
        extension_type=psk_key_exchange_modes(45), length=2
          psk_dhe_ke (1)
        extension_type=key_share(51), length=38
            NamedGroup: ecdh_x25519 (29)
            key_exchange:  (len=32): 2658F5426D3B9C89D22671D94E9A9901AF7838881C0BDA160838F37949968541

Received Record
Header:
  Version = TLS 1.0 (0x301)
  Content Type = Handshake (22)
  Length = 101
    ServerHello, Length=97
      server_version=0x301 (TLS 1.0)
      Random:
        gmt_unix_time=0x89075B1B
        random_bytes (len=28): 1EE301C35ABEB842AA67851FF14CCDD127DF78C9CBFBA57E0FC60F76
      session_id (len=32): 00AB66D0D4A897C5141A4723F712CAEB0B654627DA240C25B3B4D52967DEB286
      cipher_suite {0xC0, 0x14} TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
      compression_method: No Compression (0x00)
      extensions, length = 25
        extension_type=renegotiate(65281), length=1
            <EMPTY>
        extension_type=server_name(0), length=0
        extension_type=ec_point_formats(11), length=4
          uncompressed (0)
          ansiX962_compressed_prime (1)
          ansiX962_compressed_char2 (2)
        extension_type=encrypt_then_mac(22), length=0
        extension_type=extended_master_secret(23), length=0

SNIP  -- excluding the certificate verification which appears to be working fine

Received Record
Header:
  Version = TLS 1.0 (0x301)
  Content Type = Handshake (22)
  Length = 298
    ServerKeyExchange, Length=294
      KeyExchangeAlgorithm=ECDHE
        named_curve: ecdh_x25519 (29)
        point (len=32): 8CBCAEBEABDB43A75903B7742D564B97F2D6FEC5550E1EBC9508D6378E5D792C
      Signature (len=256): 4CDD3317DBE9ECE03E2E664D5D26B5CD62C6CEC9C790FADD26C37F35DE61D060F2AEC12258E409B9C929376A8804442CC02B8E33C831CFF64159ED2F8E9B8FAC8C295E39EE3A1903FF8C4FA524E258CD0F1786EF3777994360AB00342ACAD405267328166B7D3B54BF57E528A1C742711ABBD7C597CD733B51087284F3E592F3A95D05C2FC28FB19CE085213EC8788C0FD27FFA68EB0A429662160D0BC42DA5D78E90CB9E2AE210D9B2ACF8BCC9F69A03B9EDE0FBCC8B5C7BE1DBE2F518CE174B77F96DB0150D2A5B22B7BCF328C45693113EED12B35140FBE398C6A64FBAB62340303221C3BD183ADC75B6661A306A12C6B3DD53112F06F053A320338DB77B1

Sent Record
Header:
  Version = TLS 1.0 (0x301)
  Content Type = Alert (21)
  Length = 2
    Level=fatal(2), description=internal error(80)

603A0000:error:0A0C0103:SSL routines:tls_process_key_exchange:internal error:ssl\statem\statem_clnt.c:2254:
```

There does not appear to be an equivalent to s_client -trace in OpenSSL 1.1.1, so I used "-debug" instead and the connection was established setting up Cipher ECDHE-RSA-AES256-SHA:
```
New, TLSv1.0, Cipher is ECDHE-RSA-AES256-SHA
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1
    Cipher    : ECDHE-RSA-AES256-SHA
    Session-ID: F599F3140954DCF5E6E46CCF9436F0EB082ACE6592DC8B33B3ACF69AEF2F65FC
    Session-ID-ctx: 
    Master-Key: 7BF95BD915A8F580253F16101E728BA88FC5FC54BAC82C1ECABC72937426DC0443E341315FB84534E25EF254773AD229
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1673577058
    Timeout   : 7200 (sec)
    Verify return code: 19 (self signed certificate in certificate chain)
    Extended master secret: yes
```

**Comment by mazer1310** [comment]
Follow up:  I just found #15335 which indicates that TLS 1.0 and 1.1 were disabled in the default OpenSSL 3.0.  So, this becomes a documentation and error message request instead.

The information on the site was very opaque and the error message of "internal error" was not helpful. 

One other note.
The description of #15335 states:
```
  BCP 195 has been formally updated by RFC 8996 to deprecate TLS 1.0/1.1 and DTLS 1.0

  Ref: https://www.rfc-editor.org/info/bcp195

  In RFC parlance "TLS 1.0/1.1/DTLS 1.0 MUST NOT be used. Negotiation of TLS 1.0/1.1/DTLS 1.0 from any version of
TLS MUST NOT be permitted"
```

However, that doesn't appear to be accurate or has been since changed.  Section 3.1 of bcp195 https://www.ietf.org/rfc/bcp/bcp195.html#section-3.1.1 states:
```
  Implementations SHOULD NOT negotiate TLS version 1.0 [[RFC2246](https://www.rfc-editor.org/rfc/rfc2246)];
      the only exception is when no higher version is available in the
      negotiation.

      Rationale: TLS 1.0 (published in 1999) does not support many
      modern, strong cipher suites.  In addition, TLS 1.0 lacks a per-
      record Initialization Vector (IV) for CBC-based cipher suites and
      does not warn against common padding errors.

   o  Implementations SHOULD NOT negotiate TLS version 1.1 [[RFC4346](https://www.rfc-editor.org/rfc/rfc4346)];
      the only exception is when no higher version is available in the
      negotiation.

      Rationale: TLS 1.1 (published in 2006) is a security improvement
      over TLS 1.0 but still does not support certain stronger cipher
      suites.
```

**Comment by mattcaswell** [comment]
This looks the same as #19867. From one of my comments in that other issue:

> The problem is because we are trying to negotiate TLSv1 we are falling back to the legacy TLSv1 signature algorithm which is based on SHA1. That doesn't have enough security strength for our default security level.

The error message from OpenSSL is particularly unhelpful here. That's been fixed in 97b8db1af2f7 so should be better from OpenSSL 3.0.8 when that's released.

**Comment by t8m** [comment]
Closing as duplicate of #19867
